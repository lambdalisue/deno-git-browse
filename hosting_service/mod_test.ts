import { assertThrows } from "jsr:@std/assert@^0.221.0";
import { assertSnapshot } from "jsr:@std/testing@^0.221.0/snapshot";
import { getHostingService, UnsupportedHostingServiceError } from "./mod.ts";

Deno.test("getHostingService", async (t) => {
  await t.step("throws error for unsupported hosting service", () => {
    const url = new URL("https://example.com/lambdalisue/deno-git-browse");
    assertThrows(
      () => {
        return getHostingService(url);
      },
      UnsupportedHostingServiceError,
    );
  });

  const urls = [
    new URL("ssh://git@github.com/lambdalisue/deno-git-browse"),
    new URL("https://github.com/lambdalisue/deno-git-browse"),
    new URL("ssh://git@gitlab.com/lambdalisue/deno-git-browse"),
    new URL("https://gitlab.com/lambdalisue/deno-git-browse"),
    new URL("ssh://git@bitbucket.org/lambdalisue/deno-git-browse"),
    new URL("https://bitbucket.org/lambdalisue/deno-git-browse"),
  ];
  for (const url of urls) {
    const svc = getHostingService(url);

    await t.step(`getHomeURL for ${url}`, async () => {
      const result = await svc.getHomeURL(url);
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getCommitURL for ${url}`, async () => {
      const result = await svc.getCommitURL(url, "v0.1.0");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getTreeURL for ${url}`, async () => {
      const result = await svc.getTreeURL(url, "v0.1.0", "bin");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL for ${url} (Markdown)`, async () => {
      const result = await svc.getBlobURL(url, "v0.1.0", "README.md");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(
      `getBlobURL with lineStart for ${url} (Markdown)`,
      async () => {
        const result = await svc.getBlobURL(url, "v0.1.0", "README.md", {
          range: 10,
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    await t.step(
      `getBlobURL with lineStart/lineEnd for ${url} (Markdown)`,
      async () => {
        const result = await svc.getBlobURL(url, "v0.1.0", "README.md", {
          range: [10, 20],
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    await t.step(`getBlobURL for ${url}`, async () => {
      const result = await svc.getBlobURL(url, "v0.1.0", "bin/browse.ts");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL with lineStart for ${url}`, async () => {
      const result = await svc.getBlobURL(url, "v0.1.0", "bin/browse.ts", {
        range: 10,
      });
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(
      `getBlobURL with lineStart/lineEnd for ${url}`,
      async () => {
        const result = await svc.getBlobURL(url, "v0.1.0", "bin/browse.ts", {
          range: [10, 20],
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    if (svc.getPullRequestURL) {
      await t.step(`getPullRequestURL for ${url}`, async () => {
        const result = await svc.getPullRequestURL!(url, 1);
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      });
    }
  }
});

Deno.test("getHostingService with alias", async (t) => {
  const urls = [
    new URL("https://my-github.com/lambdalisue/deno-git-browse"),
  ];
  const aliases = {
    "my-github.com": "github.com",
  };

  for (const url of urls) {
    const svc = getHostingService(url, { aliases });

    await t.step(`getHomeURL for ${url}`, async () => {
      const result = await svc.getHomeURL(url);
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getCommitURL for ${url}`, async () => {
      const result = await svc.getCommitURL(url, "v0.1.0");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getTreeURL for ${url}`, async () => {
      const result = await svc.getTreeURL(url, "v0.1.0", "bin");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL for ${url}`, async () => {
      const result = await svc.getBlobURL(url, "v0.1.0", "README.md");
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(`getBlobURL with lineStart for ${url}`, async () => {
      const result = await svc.getBlobURL(url, "v0.1.0", "README.md", {
        range: 10,
      });
      await assertSnapshot(t, {
        url: url.href,
        result: result.href,
      });
    });

    await t.step(
      `getBlobURL with lineStart/lineEnd for ${url}`,
      async () => {
        const result = await svc.getBlobURL(url, "v0.1.0", "README.md", {
          range: [10, 20],
        });
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      },
    );

    if (svc.getPullRequestURL) {
      await t.step(`getPullRequestURL for ${url}`, async () => {
        const result = await svc.getPullRequestURL!(url, 1);
        await assertSnapshot(t, {
          url: url.href,
          result: result.href,
        });
      });
    }
  }
});
