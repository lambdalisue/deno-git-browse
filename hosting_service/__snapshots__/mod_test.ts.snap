export const snapshot = {};

snapshot[`getHostingService 1`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 2`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 3`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 4`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 5`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 6`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 7`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 8`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts?plain=1#L10",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 9`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts?plain=1#L10-L20",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 10`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/pull/1",
  url: "ssh://git@github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 11`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 12`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 13`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 14`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 15`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 16`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 17`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 18`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts?plain=1#L10",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 19`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/blob/v0.1.0/bin/browse.ts?plain=1#L10-L20",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 20`] = `
{
  result: "https://github.com/lambdalisue/deno-git-browse/pull/1",
  url: "https://github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 21`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 22`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 23`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/tree/v0.1.0/bin",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 24`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 25`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md?plain=1#L10",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 26`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md?plain=1#L10-20",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 27`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 28`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts?plain=1#L10",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 29`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts?plain=1#L10-20",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 30`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/merge_requests/1",
  url: "ssh://git@gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 31`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 32`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 33`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/tree/v0.1.0/bin",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 34`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 35`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md?plain=1#L10",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 36`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/README.md?plain=1#L10-20",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 37`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 38`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts?plain=1#L10",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 39`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/blob/v0.1.0/bin/browse.ts?plain=1#L10-20",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 40`] = `
{
  result: "https://gitlab.com/lambdalisue/deno-git-browse/-/merge_requests/1",
  url: "https://gitlab.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 41`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 42`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/commits/2120c87633f4059656e8aada8eebe62768892b46",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 43`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 44`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/README.md",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 45`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/annotate/v0.1.0/README.md#lines-10",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 46`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/annotate/v0.1.0/README.md#lines-10:20",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 47`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 48`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts#lines-10",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 49`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts#lines-10:20",
  url: "ssh://git@bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 50`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 51`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/commits/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 52`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 53`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/README.md",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 54`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/annotate/v0.1.0/README.md#lines-10",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 55`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/annotate/v0.1.0/README.md#lines-10:20",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 56`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 57`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts#lines-10",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService 58`] = `
{
  result: "https://bitbucket.org/lambdalisue/deno-git-browse/src/v0.1.0/bin/browse.ts#lines-10:20",
  url: "https://bitbucket.org/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 1`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 2`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/commit/2120c87633f4059656e8aada8eebe62768892b46",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 3`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/tree/v0.1.0/bin",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 4`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 5`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 6`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/blob/v0.1.0/README.md?plain=1#L10-L20",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;

snapshot[`getHostingService with alias 7`] = `
{
  result: "https://my-github.com/lambdalisue/deno-git-browse/pull/1",
  url: "https://my-github.com/lambdalisue/deno-git-browse",
}
`;
